File: vane-core/src/main/java/org/oddlama/vane/core/menu/Filter.java
package org.oddlama.vane.core.menu;

import java.util.List;
import java.util.stream.Collectors;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.oddlama.vane.core.functional.Function2;
import org.oddlama.vane.core.menu.Menu.ClickResult;
import org.oddlama.vane.core.module.Context;

public interface Filter<T> {
    public void openFilterSettings(
        final Context<?> context,
        final Player player,
        final String filterTitle,
        final Menu returnTo
    );

    public void reset();

    public List<T> filter(final List<T> things);

    public static class StringFilter<T> implements Filter<T> {

        private String str = null;
        private Function2<T, String, Boolean> doFilter;
        private boolean ignoreCase;

        public StringFilter(final Function2<T, String, Boolean> doFilter) {
            this(doFilter, true);
        }

        public StringFilter(final Function2<T, String, Boolean> doFilter, boolean ignoreCase) {
            this.doFilter = doFilter;
            this.ignoreCase = ignoreCase;
        }

        @Override
        public void openFilterSettings(
            final Context<?> context,
            final Player player,
            final String filterTitle,
            final Menu returnTo
        ) {
            MenuFactory.anvilStringInput(
                context,
                player,
                    filterTitle,
                new ItemStack(Material.PAPER),
                "?",
                (p, menu, s) -> {
                    menu.close(p);
                    str = s;
                    returnTo.open(p);
                    return ClickResult.SUCCESS;
                }
            ).open(player);
        }

        @Override
        public void reset() {
            str = null;
        }

        @Override
        public List<T> filter(final List<T> things) {
            if (str == null) {
                return things;
            } else {
                final String fStr;
                if (ignoreCase) {
                    fStr = str.toLowerCase();
                } else {
                    fStr = str;
                }

                return things.stream().filter(t -> doFilter.apply(t, fStr)).collect(Collectors.toList());
            }
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/menu/Filter.kt
package org.oddlama.vane.core.menu

import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.oddlama.vane.core.functional.Function2
import org.oddlama.vane.core.module.Context

/**
 * Filter contract for selector menus.
 *
 * @param T element type being filtered.
 */
interface Filter<T> {
    /**
     * Opens UI that lets the player configure filter state.
     */
    fun openFilterSettings(context: Context<*>, player: Player, filterTitle: String, returnTo: Menu?)

    /**
     * Resets filter state.
     */
    fun reset()

    /**
     * Returns a filtered copy of the provided elements.
     */
    fun filter(things: MutableList<T?>): MutableList<T?>

    /**
     * String-backed filter implementation using an anvil input prompt.
     *
     * @param T element type being filtered.
     * @param doFilter predicate evaluating whether an element matches the filter string.
     * @param ignoreCase whether filter input should be normalized to lowercase.
     */
    class StringFilter<T> @JvmOverloads constructor(
        private val doFilter: Function2<T?, String?, Boolean?>,
        private val ignoreCase: Boolean = true
    ) : Filter<T> {
        /**
         * Current filter string.
         */
        private var str: String? = null

        /**
         * Opens an anvil input menu to update the filter string.
         */
        override fun openFilterSettings(context: Context<*>, player: Player, filterTitle: String, returnTo: Menu?) {
            MenuFactory.anvilStringInput(context, player, filterTitle, ItemStack(Material.PAPER), "?") { t1, t2, t3 ->
                t2!!.close(t1!!)
                str = t3
                returnTo?.open(t1)
                Menu.ClickResult.SUCCESS
            }.open(player)
        }

        /**
         * Clears the active filter string.
         */
        override fun reset() { str = null }

        /**
         * Filters elements using the configured predicate.
         */
        override fun filter(things: MutableList<T?>): MutableList<T?> {
            val fStr = str?.let { if (ignoreCase) it.lowercase() else it } ?: return things
            return things.filterTo(mutableListOf()) { doFilter.apply(it, fStr) == true }
        }
    }
}
